File: packages/python/src/massed_compute_mcp/commands/uninstall_client.py
"""`massed-compute-mcp uninstall-client <client>` — remove our entry from a
client's MCP config. Always backs up first."""

from __future__ import annotations

import json
import os
import re
import shutil
import sys
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path
from typing import Literal

ClientId = Literal["claude-code", "cursor", "claude-desktop", "codex"]


@dataclass
class _Target:
    id: ClientId
    display_name: str
    config_path: Path
    format: Literal["json", "toml-section"]
    mcp_servers_key: str = "mcpServers"


def _claude_desktop_path() -> Path:
    home = Path.home()
    if sys.platform == "darwin":
        return home / "Library" / "Application Support" / "Claude" / "claude_desktop_config.json"
    if sys.platform == "win32":
        appdata = os.environ.get("APPDATA")
        base = Path(appdata) if appdata else home / "AppData" / "Roaming"
        return base / "Claude" / "claude_desktop_config.json"
    return home / ".config" / "Claude" / "claude_desktop_config.json"


def _targets() -> dict[ClientId, _Target]:
    home = Path.home()
    return {
        "claude-code": _Target(
            id="claude-code",
            display_name=f"Claude Code ({home / '.claude.json'})",
            config_path=home / ".claude.json",
            format="json",
        ),
        "cursor": _Target(
            id="cursor",
            display_name=f"Cursor ({home / '.cursor' / 'mcp.json'})",
            config_path=home / ".cursor" / "mcp.json",
            format="json",
        ),
        "claude-desktop": _Target(
            id="claude-desktop",
            display_name=f"Claude Desktop ({_claude_desktop_path()})",
            config_path=_claude_desktop_path(),
            format="json",
        ),
        "codex": _Target(
            id="codex",
            display_name=f"Codex ({home / '.codex' / 'config.toml'})",
            config_path=home / ".codex" / "config.toml",
            format="toml-section",
        ),
    }


SERVER_KEY = "massed-compute"
CODEX_HEADER = f"[mcp_servers.{SERVER_KEY}]"


def _backup_timestamp() -> str:
    return datetime.now().strftime("%Y%m%d-%H%M%S")


def _print_usage(targets: dict[ClientId, _Target]) -> None:
    print("Usage: massed-compute-mcp uninstall-client <client>")
    print()
    print("Supported clients:")
    for t in targets.values():
        print(f"  {t.id:<16} {t.display_name}")


def _remove_from_json(file: Path, mcp_servers_key: str) -> int:
    if not file.exists():
        print(f"{file} does not exist — nothing to remove.")
        return 0
    raw = file.read_text(encoding="utf-8")
    try:
        parsed = json.loads(raw)
    except json.JSONDecodeError:
        print(f"Refusing to edit — {file} contains invalid JSON.", file=sys.stderr)
        return 5
    if not isinstance(parsed, dict):
        print(f"Refusing to edit — {file} top level is not a JSON object.", file=sys.stderr)
        return 5
    servers = parsed.get(mcp_servers_key)
    if not isinstance(servers, dict) or SERVER_KEY not in servers:
        print(f'No "{SERVER_KEY}" entry in {file}; nothing to do.')
        return 0
    backup = file.with_suffix(file.suffix + f".bak.{_backup_timestamp()}")
    shutil.copy2(file, backup)
    print(f"Backed up {file} → {backup}")
    nextservers = {k: v for k, v in servers.items() if k != SERVER_KEY}
    nextobj = {**parsed, mcp_servers_key: nextservers}
    tmp = file.with_suffix(file.suffix + f".tmp-{os.getpid()}")
    tmp.write_text(json.dumps(nextobj, indent=2) + "\n", encoding="utf-8")
    tmp.replace(file)
    print(f'Removed "{SERVER_KEY}" entry from {file}.')
    return 0


def _remove_from_codex(file: Path) -> int:
    if not file.exists():
        print(f"{file} does not exist — nothing to remove.")
        return 0
    raw = file.read_text(encoding="utf-8")
    if CODEX_HEADER not in raw:
        print(f"No `{CODEX_HEADER}` section in {file}; nothing to do.")
        return 0
    backup = file.with_suffix(file.suffix + f".bak.{_backup_timestamp()}")
    shutil.copy2(file, backup)
    print(f"Backed up {file} → {backup}")
    out: list[str] = []
    skipping = False
    for line in raw.splitlines():
        trimmed = line.strip()
        if not skipping and trimmed == CODEX_HEADER:
            skipping = True
            continue
        if skipping and re.match(r"^\[.+\]\s*$", trimmed):
            skipping = False
        if not skipping:
            out.append(line)
    tmp = file.with_suffix(file.suffix + f".tmp-{os.getpid()}")
    tmp.write_text("\n".join(out), encoding="utf-8")
    tmp.replace(file)
    print(f"Removed `{CODEX_HEADER}` from {file}.")
    return 0


def run_uninstall_client(argv: list[str]) -> int:
    targets = _targets()
    client_arg = argv[0] if argv else None
    if not client_arg or client_arg not in targets:
        _print_usage(targets)
        return 2 if client_arg else 0
    target = targets[client_arg]  # type: ignore[index]
    print(f"Target: {target.display_name}")
    if target.format == "json":
        return _remove_from_json(target.config_path, target.mcp_servers_key)
    return _remove_from_codex(target.config_path)

File: packages/node/src/commands/uninstall-client.ts
/**
 * `massed-compute-mcp uninstall-client <client>`
 *
 * Inverse of `install-client`: removes the massed-compute entry from a
 * supported client's MCP config. Creates a timestamped backup first so
 * the change is always reversible, just like install-client.
 *
 * Does NOT delete the config file itself, even if our entry was the only
 * one — the user's other settings stay intact.
 */

import * as fs from "node:fs";
import * as os from "node:os";
import * as path from "node:path";

type ClientId = "claude-code" | "cursor" | "claude-desktop" | "codex";

interface ClientTarget {
  id: ClientId;
  displayName: string;
  configPath: string;
  format: "json" | "toml-snippet";
  mcpServersKey?: string;
}

const claudeDesktopPath = (): string => {
  const home = os.homedir();
  if (process.platform === "darwin") {
    return path.join(home, "Library", "Application Support", "Claude", "claude_desktop_config.json");
  }
  if (process.platform === "win32") {
    const appData = process.env.APPDATA ?? path.join(home, "AppData", "Roaming");
    return path.join(appData, "Claude", "claude_desktop_config.json");
  }
  return path.join(home, ".config", "Claude", "claude_desktop_config.json");
};

const targets = (): Record<ClientId, ClientTarget> => {
  const home = os.homedir();
  return {
    "claude-code": {
      id: "claude-code",
      displayName: `Claude Code (${path.join(home, ".claude.json")})`,
      configPath: path.join(home, ".claude.json"),
      format: "json",
      mcpServersKey: "mcpServers",
    },
    cursor: {
      id: "cursor",
      displayName: `Cursor (${path.join(home, ".cursor", "mcp.json")})`,
      configPath: path.join(home, ".cursor", "mcp.json"),
      format: "json",
      mcpServersKey: "mcpServers",
    },
    "claude-desktop": {
      id: "claude-desktop",
      displayName: `Claude Desktop (${claudeDesktopPath()})`,
      configPath: claudeDesktopPath(),
      format: "json",
      mcpServersKey: "mcpServers",
    },
    codex: {
      id: "codex",
      displayName: `Codex (${path.join(home, ".codex", "config.toml")})`,
      configPath: path.join(home, ".codex", "config.toml"),
      format: "toml-snippet",
    },
  };
};

interface PlainObject {
  [key: string]: unknown;
}
const isPlainObject = (v: unknown): v is PlainObject =>
  typeof v === "object" && v !== null && !Array.isArray(v);

const SERVER_KEY = "massed-compute";

const backupTimestamp = (): string => {
  const d = new Date();
  const pad = (n: number) => String(n).padStart(2, "0");
  return `${d.getFullYear()}${pad(d.getMonth() + 1)}${pad(d.getDate())}-${pad(d.getHours())}${pad(d.getMinutes())}${pad(d.getSeconds())}`;
};

const printUsage = () => {
  console.log("Usage: massed-compute-mcp uninstall-client <client>");
  console.log("");
  console.log("Supported clients:");
  for (const t of Object.values(targets())) {
    console.log(`  ${t.id.padEnd(16)} ${t.displayName}`);
  }
};

const removeFromJson = (file: string, mcpServersKey: string): number => {
  if (!fs.existsSync(file)) {
    console.log(`${file} does not exist — nothing to remove.`);
    return 0;
  }
  const raw = fs.readFileSync(file, "utf8");
  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch {
    console.error(`Refusing to edit — ${file} contains invalid JSON.`);
    return 5;
  }
  if (!isPlainObject(parsed)) {
    console.error(`Refusing to edit — ${file} top level is not a JSON object.`);
    return 5;
  }
  const servers = parsed[mcpServersKey];
  if (!isPlainObject(servers) || !(SERVER_KEY in servers)) {
    console.log(`No "${SERVER_KEY}" entry in ${file}; nothing to do.`);
    return 0;
  }
  const backup = `${file}.bak.${backupTimestamp()}`;
  fs.copyFileSync(file, backup);
  console.log(`Backed up ${file} → ${backup}`);

  const nextServers: PlainObject = { ...servers };
  delete nextServers[SERVER_KEY];
  const next: PlainObject = { ...parsed, [mcpServersKey]: nextServers };

  const tmp = `${file}.tmp-${process.pid}`;
  fs.writeFileSync(tmp, JSON.stringify(next, null, 2) + "\n", "utf8");
  fs.renameSync(tmp, file);
  console.log(`Removed "${SERVER_KEY}" entry from ${file}.`);
  return 0;
};

/**
 * Codex's config is TOML. Rather than pulling in a TOML parser (and its
 * supply-chain risk), we do a line-based removal of the
 * `[mcp_servers.massed-compute]` section and its body up to the next
 * `[...]` header. This is safe for our entries (we only ever write
 * canonical shapes) but is best-effort for hand-edited files.
 */
const removeFromCodex = (file: string): number => {
  if (!fs.existsSync(file)) {
    console.log(`${file} does not exist — nothing to remove.`);
    return 0;
  }
  const raw = fs.readFileSync(file, "utf8");
  const sectionHeader = `[mcp_servers.${SERVER_KEY}]`;
  if (!raw.includes(sectionHeader)) {
    console.log(`No \`${sectionHeader}\` section in ${file}; nothing to do.`);
    return 0;
  }
  const backup = `${file}.bak.${backupTimestamp()}`;
  fs.copyFileSync(file, backup);
  console.log(`Backed up ${file} → ${backup}`);

  const lines = raw.split(/\r?\n/);
  const out: string[] = [];
  let skipping = false;
  for (const line of lines) {
    const trimmed = line.trim();
    if (!skipping && trimmed === sectionHeader) {
      skipping = true;
      continue;
    }
    if (skipping && /^\[.+\]\s*$/.test(trimmed)) {
      skipping = false; // entered next section
    }
    if (!skipping) out.push(line);
  }
  const tmp = `${file}.tmp-${process.pid}`;
  fs.writeFileSync(tmp, out.join("\n"), "utf8");
  fs.renameSync(tmp, file);
  console.log(`Removed \`${sectionHeader}\` from ${file}.`);
  return 0;
};

export const runUninstallClient = async (argv: string[]): Promise<number> => {
  const TARGETS = targets();
  const clientArg = argv[0];
  if (!clientArg || !(clientArg in TARGETS)) {
    printUsage();
    return clientArg ? 2 : 0;
  }
  const target = TARGETS[clientArg as ClientId];
  console.log(`Target: ${target.displayName}`);
  if (target.format === "json") {
    return removeFromJson(target.configPath, target.mcpServersKey ?? "mcpServers");
  }
  return removeFromCodex(target.configPath);
};
